09-reproducible-data.Rmd

```diff
@@ -1 +1,83 @@
+
+## Version Control: Git
+
+Often, a survey analysis project produces a lot of code. Keeping track of the latest version can become challenging as code evolves throughout a project. If a team of analysts is working on the same script, someone may use an outdated version, resulting in incorrect results or duplicative work.
+
+Version control systems like Git can help alleviate these pains. Git is a system that helps track changes in computer files. Survey analysts can use Git to follow code evaluation and manage asynchronous work. With Git, it is easy to see any changes made in a script, revert changes, and resolve differences between code versions (called conflicts).
+
+Services such as GitHub or GitLab provide hosting and sharing of files as well as version control with Git. For example, we can visit the GitHub repository for this book ([https://github.com/tidy-survey-r/tidy-survey-book](https://github.com/tidy-survey-r/tidy-survey-book)) and see the files that build the book, when they were committed to the repository, and the history of modifications over time.
+
+<!--TODO: Add image-->
+
+In addition to code scripts, platforms like GitHub can store data and documentation. They provide a way to maintain a history of data modifications through versioning and timestamps. By saving the data and documentation alongside the code, it becomes easier for others to refer to and access everything they need in one place.
+
+Using version control in analysis projects makes collaboration and maintenance more manageable. For connecting Git with R, we recommend the [Happy Git and GitHub for the useR by Jenny Bryan and Jim Hester](https://happygitwithr.com/) [@git-w-R].
+
+## Package Management: {renv}
+
+Along with version control of code, version control for packages is also important for reproducibility.  If two people run the same code but use different versions of a package, the outcomes could be different depending on the package updates.  For example, this book currently uses a version of the {srvyr} package from GitHub and not from CRAN.  This is because the version of {srvyr} has some bugs (errors) when doing some calculations. The version on GitHub has corrected these errors, so we have asked users to install the GitHub version to obtain the same results. 
+
+One way to easily handle different package versions is with the {renv} package. This package allows researchers to set the versions for each package used and easily manage package dependencies. Specifically, {renv} creates isolated, project-specific environments that record the packages and their versions used in the code. When initiated by a new user, {renv} checks whether the installed packages are consistent with the recorded version for the project. If not, it installs the correct versions and ensures that others can replicate the project's environment to rerun the code and obtain consistent results.
```

Comment:
I had written this a while ago, but never finished:

### Project-based workflows

We recommend a project-based workflow for analysis projects as described in Hadley Wickham Mine Çetinkaya-Rundel, and Garrett Grolemund's book, R for Data Science, found at [r4ds.hadley.nz/](https://r4ds.hadley.nz/). A project-based workflow helps us with file system discipline, since we put everything related to a project in a designated folder. Since all associated files are in a single location, they are easy to find and organize.

The RStudio IDE has built-in support for projects. When we create a project in RStudio, it creates a `.Rproj` file that store settings specific to that project. Once we have created a project, we can create folders that help us organize our workflow. For example, a project directory could look like this:

```
| anes_analysis/
  | anes_analysis.Rproj
  | README.md
  | codebooks
    | codebook2020.pdf
    | codebook2016.pdf
  | rawdata
    | anes2020_raw.csv
    | anes2016_raw.csv
  | scripts
    | data-prep.R
  | data
    | anes2020_clean.csv
    | anes2016_clean.csv
  | report
    | anes_report.Rmd
    | anes_report.html
    | anes_report.pdf
```

The {here} package enables easy file referencing. In a project-based workflow, all paths are relative and, by default, relative to the project’s folder. Use the `here::here()` function to build the path when we load or save data. Below, we ask R to read the CSV file `anes_2020.csv` in the project directory's `Data` folder: 

```{r}
#| eval: false
anes <- 
  read_csv(here::here("data", "anes2020_clean.csv"))
```